lib/spack/spack/test/patch.py

```diff
@@ -328,9 +328,25 @@ def test_write_and_read_sub_dags_with_patched_deps(mock_packages, config):
         spec.package.package_dir)
 
 
-def test_file_patch_no_file():
+def test_patch_no_file():
+    # Give it the attributes we need to construct the error message
+    FakePackage = collections.namedtuple(
+        'FakePackage', ['name', 'namespace', 'fullname'])
+    fp = FakePackage('fake-package', 'test', 'fake-package')
+    with pytest.raises(ValueError, match='FilePatch:'):
+        spack.patch.FilePatch(fp, 'nonexistent_file', 0, '')
```

Comment:
It seems we are testing two things here. What is the aim of the test? If `FilePatch` can be regarded as an implementation detail of a higher level API, it would be better to test it indirectly (since test that check implementation details are not very resistant to refactoring, i.e. they need a lot of change too when the implementation changes)

Reply:
> What is the aim of the test?

This test asserts that the use of the `patch` directive with a non-existing file will fail.

> If FilePatch can be regarded as an implementation detail of a higher level API, it would be better to test it indirectly (since test that check implementation details are not very resistant to refactoring, i.e. they need a lot of change too when the implementation changes)

`Patch` is essentially an ABC, with implementations in `FilePatch` and `UrlPatch`. The reason I modified this test was that `FilePatch` has its own file existence check, so the one in `Patch` never gets hit in our tests. Arguably, it would be better to refactor our patching logic, but that's outside the scope of this PR. I'm just getting us 2 more lines of coverage in this file.